### PR TITLE
Additional Net RPC API & CLI Methods

### LINKED ISSUE #881: Implement Network RPC API
**Issue summary**

Implement the JSONRPC methods pertaining to the Network API.
- [x] Filecoin.NetAddrsListen
- [x] Filecoin.NetPeers
- [x] Filecoin.NetConnect

There are a few more, but I am not sure how relevant they are. These ones seem the most important and useful.
**Other information and links**

NetDisconnect implementation broken out into #1181.

### PR DESCRIPTION
**Pull request by cryptoquick**
**Summary of changes**
Changes introduced in this pull request:
- [x] Adds NetPeers, NetConnect, and NetDisconnect methods to RPC API
- [x] Adds functionality for:
  - [x] NetPeers
  - [x] NetConnect
  - [ ] ~~NetDisconnect~~ - Broken out into #1181 
- [x] Adds JSON-RPC client and server APIs
- [x] Adds CLI commands for `net listen`, `net peers`, `net connect`, `net disconnect`

**Reference issue to close (if applicable)**

Closes #881 

**Other information and links**

### CODE REVIEW

File: forest/src/cli/net_cmd.rs [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+                        .collect();
+
+                    print_stdout(output.join("\n"));
+                }
+                Err(e) => handle_rpc_err(e),
+            },
+            Self::Connect { address } => {
+                let addr: Multiaddr = address
+                    .parse()
+                    .expect("parse provided multiaddr from string");
```
Comment (olibearo) [inline forest/src/cli/net_cmd.rs:69]: This would be a better devx as an error printed to stderr

Comment (olibearo) [reply forest/src/cli/net_cmd.rs:69]: And then an exit with the proper exit code

Comment (cryptoquick) [reply forest/src/cli/net_cmd.rs:69]: Good idea! Will do.

File: node/rpc/src/net_api.rs [hunk 2]
```diff
@@ -22,10 +23,88 @@ pub(crate) async fn net_addrs_listen<
...
+    DB: BlockStore + Send + Sync + 'static,
+    B: Beacon + Send + Sync + 'static,
+>(
+    data: Data<RPCState<DB, B>>,
+    Params(params): Params<NetDisconnectParams>,
+) -> Result<NetDisconnectResult, JsonRpcError> {
+    let (id,) = params;
+    let peer_id = PeerId::from_bytes(id.as_bytes())?;
+
+    println!("compare peer_ids: {} == {}", id, &peer_id.to_base58()); // TODO: remove
```
Comment (ec2) [inline node/rpc/src/net_api.rs:99]: Remove

File: node/rpc/src/net_api.rs [hunk 3]
```diff
@@ -22,10 +23,88 @@ pub(crate) async fn net_addrs_listen<
...
+    let (tx, rx) = oneshot::channel();
+    let req = NetworkMessage::JSONRPCRequest {
+        method: NetRPCMethods::NetConnect(tx, peer_id, addrs),
+    };
+
+    data.network_send.send(req).await?;
+    let success = rx.await?;
+
+    if success {
+        info!("connect {} success", peer_id.to_base58()); // TODO: return this from CLI
```
Comment (ec2) [inline node/rpc/src/net_api.rs:81]: Dont print here

File: node/forest_libp2p/src/service.rs [hunk 4]
```diff
@@ -321,12 +326,44 @@ where
...
+                                        };
+                                    }
+
+                                    if response_channel.send(success).is_err() {
+                                        warn!("Failed to connect to a peer");
+                                    }
+                                }
+                                NetRPCMethods::NetDisconnect(response_channel, peer_id) => {
+                                    // swarm_stream.get_mut().disconnect(&peer_id);
+                                    println!("TODO: {}", peer_id);
```
Comment (ec2) [inline node/forest_libp2p/src/service.rs:361]: Have a more informative message here

File: node/rpc-client/src/lib.rs [hunk 5]
```diff
@@ -1,12 +1,180 @@
...
+struct URL {
+    protocol: String,
+    port: String,
+    host: String,
+}
+
+/// Parses a multiaddress into a URL
+fn multiaddress_to_url(ma_str: String) -> String {
+    // Parse Multiaddress string
+    let ma: Multiaddr = ma_str.parse().expect("Parse multiaddress");
```
Comment (ec2) [inline node/rpc-client/src/lib.rs:57]: Why not return an error here?

File: node/rpc-client/src/lib.rs [hunk 6]
```diff
@@ -1,12 +1,180 @@
...
+            addr
+        },
+    );
+
+    // Format, print and return the URL
+    let url = format!(
+        "{}://{}:{}/{}",
+        addr.protocol, addr.host, addr.port, RPC_ENDPOINT
+    );
+    debug!("Using JSON-RPC v2 HTTP URL: {}", url);
```
Comment (ec2) [inline node/rpc-client/src/lib.rs:106]: Should debug log this where this method actually gets called.

File: node/forest_libp2p/src/service.rs [hunk 7]
```diff
@@ -321,12 +326,44 @@ where
...
+                                }
+                                NetRPCMethods::NetConnect(response_channel, peer_id, addresses) => {
+                                    let mut addresses = addresses.clone();
+                                    let mut success = false;
+
+                                    for multiaddr in addresses.iter_mut() {
+                                        multiaddr.push(Protocol::P2p(Multihash::from_bytes(&peer_id.to_bytes()).unwrap()));
+
+                                        if Swarm::dial_addr(swarm_stream.get_mut(), multiaddr.clone()).is_ok() {
+                                            success = true;
```
Comment (ec2) [inline node/forest_libp2p/src/service.rs:351]: Should early return if a connection is successful, no? Why try other addresses?

**Review by ec2 — CHANGES_REQUESTED** [review]

File: node/rpc-client/src/lib.rs (hunk 5 above)
Comment (cryptoquick) [reply node/rpc-client/src/lib.rs:57]: Actually, this isn't new code, this is moved code. Also, I'm completely refactoring how Multiaddrs are passed and parsed in the CLI in this PR:
https://github.com/ChainSafe/forest/pull/1183/files

File: node/rpc-client/src/lib.rs (hunk 6 above)
Comment (cryptoquick) [reply node/rpc-client/src/lib.rs:106]: How would you suggest I do that?

File: node/forest_libp2p/src/service.rs (hunk 7 above)
Comment (cryptoquick) [reply node/forest_libp2p/src/service.rs:351]: Good thinking! Will do.

File: node/forest_libp2p/src/service.rs [hunk 8]
```diff
@@ -321,12 +326,44 @@ where
...
+                                    if response_channel.send(peer_addresses.to_owned()).is_err() {
+                                        warn!("Failed to get Libp2p peers");
+                                    }
+                                }
+                                NetRPCMethods::NetConnect(response_channel, peer_id, addresses) => {
+                                    let mut addresses = addresses.clone();
+                                    let mut success = false;
+
+                                    for multiaddr in addresses.iter_mut() {
+                                        multiaddr.push(Protocol::P2p(Multihash::from_bytes(&peer_id.to_bytes()).unwrap()));
```
Comment (olibearo) [inline node/forest_libp2p/src/service.rs:348]: Does this need a `cargo fmt`? Hopefully the CI will say if so

**Review by olibearo — CHANGES_REQUESTED** [review]
Thanks for this contribution, I'm more familiar with our cli/rpc/libp2p scaffolding after reading this.

I left a couple of minor suggestions/requests, let me know what you think.

File: node/forest_libp2p/src/service.rs (hunk 8 above)
Comment (cryptoquick) [reply node/forest_libp2p/src/service.rs:348]: No, `cargo fmt` suggests no changes.

Comment (olibearo) [reply node/forest_libp2p/src/service.rs:348]: Yeah, seems like it was a Github rendering issue, not an issue with the actual file

Comment (cryptoquick) [reply node/forest_libp2p/src/service.rs:348]: Gotcha

**Review by ec2 — APPROVED** [review]

File: forest/src/cli/net_cmd.rs [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+                }
+                Err(e) => handle_rpc_err(e),
+            },
+            Self::Connect { address } => {
+                let addr: Multiaddr = address
+                    .parse()
+                    .map_err(|e| {
+                        cli_error_and_die(
+                            &format!("Error parsing multiaddr. Error was: {}", e),
+                            400,
```
Comment (connormullett) [inline forest/src/cli/net_cmd.rs:72]: The intention behind `cli_error_and_die` was for CLI specific errors, ie bad permissions, argument errors, etc. That is correct here but I don't think we should be using HTTP status codes for this. While this is a `Bad Request`, I feel this should be a *Nix error with an exit code of 1.

**Review by connormullett — CHANGES_REQUESTED** [review]
Nice job! I have one little nit that I think is warranted; Otherwise, looks good

File: forest/src/cli/net_cmd.rs (hunk 9 above)
Comment (cryptoquick) [reply forest/src/cli/net_cmd.rs]: Fixed

**Review by olibearo — APPROVED** [review]
Factoring looks good, nice job @cryptoquick

File: forest/src/cli/net_cmd.rs [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+                for protocol in addr.iter() {
+                    if let Protocol::P2p(p2p) = protocol {
+                        id = multibase::encode(multibase::Base::Base58Btc, p2p.to_bytes());
+                        id = id.split_off(1);
+                    }
+                }
+
+                if id.is_empty() {
+                    cli_error_and_die("Needs a /p2p/ protocol present in multiaddr", 400);
```
Comment (connormullett) [inline forest/src/cli/net_cmd.rs:84]: My bad, forgot to mark this one as well. This should be an exit code `1` instead of `400`

Comment (cryptoquick) [reply forest/src/cli/net_cmd.rs]: Fixed. I was working on another issue, so I overlooked it. Multitasking.

**Review by connormullett — APPROVED** [review]
LGTM! Let's get this in

**Pull request merged**